src/Hosting/Hosting/src/WebHostBuilderExtensions.cs

```diff
@@ -67,7 +71,7 @@ private static IWebHostBuilder Configure(this IWebHostBuilder hostBuilder, Actio
         /// <param name="hostBuilder">The <see cref="IWebHostBuilder"/> to configure.</param>
         /// <param name="startupFactory">A delegate that specifies a factory for the startup class.</param>
         /// <returns>The <see cref="IWebHostBuilder"/>.</returns>
-        public static IWebHostBuilder UseStartup(this IWebHostBuilder hostBuilder, Func<WebHostBuilderContext, object> startupFactory)
+        public static IWebHostBuilder UseStartup<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]TStartup>(this IWebHostBuilder hostBuilder, Func<WebHostBuilderContext, TStartup> startupFactory)
```

Comment:
What happens if someone does this:

`host.UseStartup<object>(c => return MyComplexType());`

We'll be preserving public methods on object, not MyComplexType. Should we validate the return type from the func is the same as the specified type?

Reply:
Yes this is a tricky one and I agree.